### PR TITLE
NODE-1580: Proofs ser/de

### PR DESCRIPTION
**Pull request by disfenoid**
[JIRA: https://wavesplatform.atlassian.net/browse/NODE-1580](https://wavesplatform.atlassian.net/browse/NODE-1580)

### CODE REVIEW

File: src/main/scala/com/wavesplatform/serialization/Deser.scala [hunk 1]
```diff
@@ -6,7 +6,10 @@ object Deser {
 
   def serializeBoolean(b: Boolean): Array[Byte] = if (b) Array(1: Byte) else Array(0: Byte)
 
-  def serializeArray(b: Array[Byte]): Array[Byte] = Shorts.toByteArray(b.length.toShort) ++ b
+  def serializeArray(b: Array[Byte]): Array[Byte] = {
+    val lengthBytes = Shorts.toByteArray(b.length.toShort)
```
Comment (Karasiq) [inline src/main/scala/com/wavesplatform/serialization/Deser.scala]: There should be an overflow check
```suggestion
    val lengthBytes = Shorts.toByteArray(b.length.ensuring(_.isValidShort).toShort)
```

File: src/main/scala/com/wavesplatform/serialization/Deser.scala [hunk 2]
```diff
@@ -39,5 +42,10 @@ object Deser {
 
   def serializeOption[T](b: Option[T])(ser: T => Array[Byte]): Array[Byte] = b.map(a => (1: Byte) +: serializeArray(ser(a))).getOrElse(Array(0: Byte))
 
-  def serializeArrays(bs: Seq[Array[Byte]]): Array[Byte] = Shorts.toByteArray(bs.length.toShort) ++ Bytes.concat(bs.map(serializeArray): _*)
+  def serializeArrays(bs: Seq[Array[Byte]]): Array[Byte] = {
+    val countBytes  = Shorts.toByteArray(bs.length.toShort)
```
Comment (Karasiq) [inline src/main/scala/com/wavesplatform/serialization/Deser.scala]: Same here

File: src/main/scala/com/wavesplatform/serialization/Deser.scala (hunk 1 above)
Comment (disfenoid) [reply src/main/scala/com/wavesplatform/serialization/Deser.scala]: fixed

File: src/main/scala/com/wavesplatform/serialization/Deser.scala (hunk 2 above)
Comment (disfenoid) [reply src/main/scala/com/wavesplatform/serialization/Deser.scala]: fixed

**Comment by Karasiq** [comment]
Fix test pls
```diff
Index: it/src/test/scala/com/wavesplatform/it/sync/transactions/DataTransactionSuite.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- it/src/test/scala/com/wavesplatform/it/sync/transactions/DataTransactionSuite.scala	(revision e14e21b2fda79344bdbe9ca0291036463e5c4113)
+++ it/src/test/scala/com/wavesplatform/it/sync/transactions/DataTransactionSuite.scala	(date 1556107167000)
@@ -238,7 +238,7 @@
                                 "Duplicate keys found")
 
     val extraValueData = List(BinaryDataEntry("key", ByteStr(Array.fill(MaxValueSize + 1)(1.toByte))))
-    assertBadRequestAndResponse(sender.putData(firstAddress, extraValueData, calcDataFee(extraValueData)), TooBig)
+    assertBadRequestAndResponse(sender.putData(firstAddress, extraValueData, 1.waves), TooBig)
     nodes.waitForHeightArise()
 
     val largeBinData = List.tabulate(5)(n => BinaryDataEntry(extraKey, ByteStr(Array.fill(MaxValueSize)(n.toByte))))

```

**Pull request merged**